platforms/core-configuration/model-core/src/main/java/org/gradle/api/internal/provider/CollectionSupplier.java

```diff
@@ -18,10 +18,29 @@
 
 import java.util.Collection;
 
+/**
+ * A composable {@link ValueSupplier value supplier} that supplies collection values. Instances of
+ * this type are used as value suppliers for {@link AbstractCollectionProperty}.
+ *
+ * @param <T> the type of elements
+ * @param <C> type of collection
+ */
 interface CollectionSupplier<T, C extends Collection<? extends T>> extends ValueSupplier {
+    /**
+     * Realizes and returns the (collection) value.
+     */
     Value<? extends C> calculateValue(ValueConsumer consumer);
 
-    CollectionSupplier<T, C> plus(Collector<T> collector);
+    /**
+     * Composes the current collection supplier with the given {@link Collector} (also a {@link ValueSupplier value supplier}), producing a new
```

Comment:
🤔 The bit about composability seems more like `@implSpec` here.  I don't think the interface can/should promise composition - it should merely describe the requirements for the result.